### ISSUE TITLE
Support for "pinning" to a certain version of chisel-releases

### ISSUE THREAD

**Issue by lbussell** [canonical/chisel#172]
Chisel looks at [canonical/chisel-releases](https://github.com/canonical/chisel-releases) by default to get slice definitions.

Ubuntu releases themselves are pretty stable and don't contain breaking changes, but that doesn't hold for the slice definitions. For a single release, e.g. 24.04, slice definitions can change at any time. Recent example: https://github.com/canonical/chisel-releases/commit/3bf3f9ef7429c56e3716066e7023a45858e67af0. New slices can be added, slice contents can change, etc.

When "Chiseling" Ubuntu images, sometimes we'd like to re-build an image to pick up package updates or security fixes, but without the chance of a slice changing out from under us without notice.

How can we "pin" images to a set of slice definitions and only take new slice updates when we're ready? Is there a solution that could work by cloning a specific commit of chisel-releases and using it as a local set of slice definitions?

Thanks in advance!

**Comment by cjdcordeiro** [comment]
Hi @lbussell 

This is a good point of discussion.
Our current approach to maintaining an already released chisel-release is that we don't allow regressions but we may approve additions or fixes to existing slices.

So, for example:
 - in canonical/chisel-releases@3bf3f9e , that's just refactoring. If you were using `openjdk-8-jre-headless_core`, we'd have had no changes in the content of that slice
 - https://github.com/canonical/chisel-releases/pull/306/commits/aaf916a24d27f102392fa5e2033435efba8bb620#diff-964aee46c3f2ce19718352a4077280431cd4d2c85195063bf3050801b1c8cc94L33-L55 - here, we're not allowing the deletion of an existing slice
 - in https://github.com/canonical/chisel-releases/pull/231 we are adding new content to slices. In this particular case, it is a very welcome improvement (the copyright file) despite the fact that you'll get some additional content that you didn't get before.

So in short, removal of slice contents (or the entire slice itself) shouldn't normally happen, and if it does, please raise an issue. Complementing an existing slice might happen, especially if the additions are fixing an existing problem.

If you really need to pin a certain chisel-releases commit, then you can use a local version of it. E.g.:

```bash
git clone -b ubuntu-24.10 https://github.com/canonical/chisel-releases
cd chisel-releases
git reset --hard <sha>
chisel cut --release ./ --root ....
```

I hope this helps.

(loosely related to https://github.com/canonical/chisel/issues/154)

**Comment by lbussell** [comment]
Thanks for those details!

The motivation for opening this issue is that we have some package baseline tests that validate we install the correct packages and don't accidentally add any new ones. We've been tripped up a couple times when "new packages" got added as part of some of the copyright slice additions. However, it seems like that has mostly stabilized.

I didn't see any discussion about pinning to a version of this repo, so I thought I'd just raise that idea here. Thanks for the confirmation about the slice maintenance approach and the local repo workaround! It seems like we shouldn't be _too_ concerned about any breaking changes.

**Comment by cjdcordeiro** [comment]
Hi !

Got it. 

> We've been tripped up a couple times when "new packages" got added as part of some of the copyright slice additions. However, it seems like that has mostly stabilized.

I see. In this particular case, it was indeed an intentional UX improvement. But I think we should still expect that sometimes a certain slice may get extended, simply because 1) the upstream deb may have received a patch that carries new contents, or 2) the slice is incomplete. The latter should happen less frequently now that we've introduced functional testing for slices.

Thanks for the suggestion though. If that's ok, I think I'll just cross reference this issue in https://github.com/canonical/chisel/issues/154 and close it for now ;)

Thanks again
